org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/SaveActionHandler.java

```diff
@@ -155,4 +167,26 @@ private List<TextEdit> handleSaveActionOrganizeImports(String documentUri, IProg
 		return edit;
 	}
 
+	private void handleSaveActionRenameFile(String documentUri, IProgressMonitor monitor) {
+		ICompilationUnit cu = JDTUtils.resolveCompilationUnit(documentUri);
+		CompilationUnit astRoot = CoreASTProvider.getInstance().getAST(cu, CoreASTProvider.WAIT_YES, null);
+		IProblem[] problems = astRoot.getProblems();
+		Optional<IProblem> desiredProblem = Arrays.stream(problems).filter(p -> p.getID() == IProblem.PublicClassMustMatchFileName).findFirst();
+		if (desiredProblem.isPresent()) {
+			IProblem renameProblem = desiredProblem.get();
+			String newName = renameProblem.getArguments()[1];
+			String oldName = cu.getElementName();
+			String newUri = documentUri.replace(oldName, newName + ".java");
+			List<Either<TextDocumentEdit, ResourceOperation>> documentChanges = new ArrayList<>();
+			RenameFile operation = new RenameFile(documentUri, newUri);
+			documentChanges.add(Either.forRight(operation));
+			WorkspaceEdit edit = new WorkspaceEdit(documentChanges);
+			edit.setChanges(Collections.emptyMap());
```

Comment:
Is this required ? According to the spec a [`WorkspaceEdit.changes`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceEdit) is optional.